code/game/machinery/hologram.dm

```diff
@@ -41,313 +37,275 @@ var/const/HOLOPAD_MODE = RANGE_BASED
 		master.show_message(rendered, type)
-	if(targetpad)
-		for(var/mob/living/carbon/master in view(targetpad))
-			var/rendered = "<i><span class='game say'>The holographic image of <span class='message'>[msg]</span></span></i>"
-			master.show_message(rendered, type)
-
-/obj/machinery/hologram/holopad/proc/create_holo(mob/living/silicon/ai/A, mob/living/carbon/caller_id, turf/T = loc)
-	if(hacked == 0)
-		var/obj/effect/overlay/hologram = new(T)//Spawn a blank effect at the location.
-		hologram.no_clean = TRUE
-		if(caller_id)
-			var/tempicon = getFlatIcon(caller_id)
-
-			hologram.name = "[caller_id.name] (Hologram)"
-			hologram.forceMove(get_step(src,1))
-			masters[caller_id] = hologram
-			hologram.icon = getHologramIcon(icon(tempicon)) // Add the callers image as an overlay to keep coloration!
-		else
-			hologram.icon = A.holo_icon // Add the AI's configured holo Icon
-			hologram.name = "[A.name] (Hologram)"//If someone decides to right click.
-			A.holo = src
-			masters[A] = hologram
-
-		hologram.layer = FLY_LAYER//Above all the other objects/mobs. Or the vast majority of them.
-		hologram.mouse_opacity = 0//So you can't click on it.
-		hologram.anchored = 1//So space wind cannot drag it.
-		hologram.set_light(2)	//hologram lighting
-		hologram.color = color //painted holopad gives coloured holograms
-		set_light(2)			//pad lighting
-		icon_state = "holopad1"
-		return 1
-	else if(hacked == 1)
-		var/obj/effect/overlay/hologram = new(T)//Spawn a blank effect at the location.
-		hologram.no_clean = TRUE
-		hologram.mouse_opacity = 0//So you can't click on it.
-		hologram.anchored = 1//So space wind cannot drag it.
-		masters[A] = hologram
-		A.holo = src
-		return 1
-
-/obj/machinery/hologram/holopad/proc/clear_holo(mob/living/silicon/ai/user, mob/living/carbon/call_user)
-	if(user)
-		qdel(masters[user])//Get rid of user's hologram
-		user.holo = null
-		masters -= user //Discard AI from the list of those who use holopad
-	if(call_user)
-		qdel(masters[call_user]) //Get rid of user's hologram
-		masters -= call_user //Discard the caller from the list of those who use holopad
-	if (!masters.len)//If no users left
-		set_light(0)			//pad lighting (hologram lighting will be handled automatically since its owner was deleted)
-		icon_state = "holopad0"
-		if(sourcepad)
-			sourcepad.targetpad = null
-			sourcepad = null
-			caller_id = null
-	return 1
 
-/obj/machinery/hologram/holopad/machinery_process()
-	for (var/mob/living/silicon/ai/master in masters)
-		var/active_ai = (master && !master.incapacitated() && master.client && master.eyeobj)//If there is an AI with an eye attached, it's not incapacitated, and it has a client
-		if((stat & NOPOWER) || !active_ai)
-			clear_holo(master)
+/obj/machinery/hologram/holopad/proc/create_holos()
+	for(var/mob/living/M in view(world.view, connected_pad))
+		if(LAZYISIN(active_holograms, M))
+			continue
+		create_holo(M)
+
+/obj/machinery/hologram/holopad/proc/create_holo(mob/M)
```

Comment:
Should we not have a check in here to prevent the holos from being created in an area the holocomm cannot see? e.g. in a wall, in a different room, etc?

Reply:
it looks a bit janky, but i think holotech should look a bit janky, clipping through walls and stuff, because you're not actually there. i also dunno if it's worth the performance to check constantly, but lemme know what you think

Reply:
It only needs to be checked when:
- The hologram is created
- The someone moves near the holopad

So I don't think it's too bad.

You can use some line of sight calculation rather than doing view() potentially? I think we have helpers for that.

Reply:
You could also make the holopad update holograms less often for a 'janky' effect, if you want? Holograms teleporting around a bit rather than walking smoothly.

Reply:
currently the holograms are updated on machinery_process, because stuff needs to be updated to account for the various things that can happen to mobs (resting, taking off clothes, changing direction, moving). if you can make a branch with all that stuff, feel free to merge it into here